### ISSUE TITLE
Cache manager with custom classloader produces ClassNotFoundException when deserialize cache entry (OSGi env)

### ISSUE THREAD

**Issue by exmortal** [hazelcast/hazelcast#5761]
Hi,

I found a strange situation with a cache manager classloader in OSGi environment (Apache KARAF container)

I have 2 bundles. One of them contains hazelcast jar and takes care about cache manager initialization. (create and expose it via OSGi service method). Second one needs to get a cache instance from provider initialized in bundle1. Also it contains CacheItem class that should be used as a value type.

Here is a test code: 
```java
// get an instance of JSR107 cache provider created in bundle1
CachingProvider provider = bundle1CacheProviderService.getProvider();
// get an instance of cache manager for the bundle2 classloader
CacheManager manager = provider.getCacheManager(bundle2ClassLoader, null);

// create cache instance with empty configuration
MutableConfiguration configuration = new MutableConfiguration<>();
final Cache cache = manager.createCache(cacheName, configuration);

CacheItem item1 = new CacheItem();
cache.put("1", item1); // OK
cache.get("1"); // ClassNotFoundException. 
```
Dezerializer uses context classloader but not one I specified when create a cache manager.

I have a quick look to the source code and found that manager classloader is used only for manager's identification by pair URI + classloader. Cache instances use a context classloader for deserialization.

I know that if I use a own cache provider in each bundle will solve this issue, but I think that it will be a real overheat.

I just want to know is it a bug or expected behavior.

Thanks,

BR,
Vlad

**Comment by serkan-ozal** [comment]
Hi @exmortal,

Sorry for late response.

At first look, it seems as an issue to me but I need to look more detailed. 

When the default URI (`hazelcast`) is used (no URI is specified while creating `CacheManager`), specified classloader is not passed to the underlying instance but passed when URI is specified while creating cache manager. 

So to be sure, can you pass a non-empty (null) URI to `CachingProvider` to create a `CacheManager` that uses a `HazelcastInstance` with the specified classloader at background?

**Comment by exmortal** [comment]
Hi @serkan-ozal 

It seems that nothing changes either I set URI to something of leave it as null. I still get the deserialization error. 

BR, 
Vlad

**Comment by serkan-ozal** [comment]
Hi @exmortal,

Can you send the stacktrace?
Does the issue occurred while casting the returned object from `cache.get()` to a local variable like `CacheItem item = cache.get("1");` or just `cache.get("1");` itself causes `ClassCastException`?

**Comment by exmortal** [comment]
java.lang.ClassNotFoundException: com.sample.cache.CacheItem not found by org.ops4j.pax.exam.rbc [53]
        at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1532)
        at org.apache.felix.framework.BundleWiringImpl.access$400(BundleWiringImpl.java:75)
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1955)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:356)
        at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:125)
        at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:114)
        at com.hazelcast.nio.IOUtil$1.resolveClass(IOUtil.java:113)
        at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1593)
        at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1514)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1750)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1347)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:369)
        at com.hazelcast.nio.serialization.DefaultSerializers$ObjectSerializer.read(DefaultSerializers.java:196)
        at com.hazelcast.nio.serialization.StreamSerializerAdapter.read(StreamSerializerAdapter.java:41)
        at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:276)
        at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:200)
        at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveApplicationResponse(InvocationFuture.java:375)
        at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveApplicationResponseOrThrowException(InvocationFuture.java:334)
        at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:225)
        at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:204)
        at com.hazelcast.cache.impl.AbstractCacheProxy.get(AbstractCacheProxy.java:158)
        at com.hazelcast.cache.impl.CacheProxy.get(CacheProxy.java:71)
        at com.hazelcast.cache.impl.CacheProxy.get(CacheProxy.java:87)
        at com.sample.cache.HazelcastCacheJsr107IT.testCreateCacheInCorrectClassLoader(HazelcastCacheJsr107IT.java:103)

**Comment by serkan-ozal** [comment]
Ok. It is `ClassNotFoundException` not `ClassCastException`. I am a little bit confused because the title of the issue says `ClassNotFoundException` but the comment in the code sample says `ClassCastException` (`cache.get("1"); // ClassCastException. Dezerializer uses context classloader but not one I specified when create a cache manager.`)

**Comment by exmortal** [comment]
Yes, sorry, It's my fault. I've just updated first message.

**Comment by serkan-ozal** [comment]
Are you sure that `com.sample.cache.CacheItem` can be loaded/found by `bundle2ClassLoader`?
Can you be sure by trying `bundle2ClassLoader.loadClass("com.sample.cache.CacheItem")` doesn't throw `ClassNotFoundException`?

**Comment by exmortal** [comment]
Yes, sure. Class is loaded via bundle2ClassLoader.

**Comment by exmortal** [comment]
Moreover: this code fix the problem:

'''java
CacheItem item2 = doWithClassLoader(bundle2ClassLoader, new Callable<CacheItem>() {
            @Override
            public CacheItem call() throws Exception {
                return cache.get("1");
            }
        });

 just set a correct classloader as a actual thread context one.

**Comment by serkan-ozal** [comment]
Ok thanks. This is very helpful and proves that context classloader is used internally.

**Comment by serkan-ozal** [comment]
@exmortal Can you send the version of the code that you have added URI?

**Comment by serkan-ozal** [comment]
@exmortal 

BTW, Can you also print the classloader (used while serialization) of the underlying Hazelcast instance by `CacheManager`
``` java
HazelcastInstanceImpl instanceImpl =
                (HazelcastInstanceImpl) (((HazelcastCacheManager) manager).getHazelcastInstance());
SerializationService ss = instanceImpl.getSerializationService();
ClassLoader cl = ss.getClassLoader();
System.out.println(cl);
```

**Comment by exmortal** [comment]
URI is "simple/cache".

It's unable to invoke your part of code. I've got: 
```java
java.lang.ClassCastException: com.hazelcast.instance.HazelcastInstanceProxy cannot be cast to com.hazelcast.instance.HazelcastInstanceImpl
```
But debug shows me that classloader is null in SerializationService.

BTW I create a cache manager in this way:
```java
Properties props = new Properties();
props.putAll(properties);
props.setProperty(HazelcastCachingProvider.HAZELCAST_INSTANCE_NAME, hazelcastInstance.getName());

URI uri = new URI("simple/cache");
CacheManager cacheManager = provider.getCacheManager(uri, classLoader, props);
```
Cache manager uses already created and configured hazelcast instance in my case.

**Comment by serkan-ozal** [comment]
I see,

Can you set the `bundle2ClassLoader` to `Config` by

``` java
Config config = new Config();
config.setClassLoader(bundle2ClassLoader);
...
```

and then create Hazelcast instance with this config?

**Comment by serkan-ozal** [comment]
In your example, you are creating an instance and I guess you don't give the required classloader to it. Although you give the classloader while creating a cache, it does't create a new instance but used an existing instance with its name but this instance has no classloader and uses thread context classloader in its serialization.

**Comment by exmortal** [comment]
I can't set bundle2ClassLoader while hazelcast initialization, because hazelcast instance must be created in bundle1 by design in my particular case. Bundle1 is a cache producer. Bundle2 is a cache consumer. Bundle2 retrieves a cache manager only, but doesn't create a new cache provider.

**Comment by exmortal** [comment]
I can set only bundle1Classloader to the config, but I believe this won't help.

**Comment by serkan-ozal** [comment]
In your architecture, you are using an Hazelcast instance from another bundle but that bundle doesn't know the class from this bundle as expected. I don't think this is a Hazelcast bug. If you want Hazelcast to find your class, you should give your custom classloader (maybe a global classloader that scans the requested class in all bundles) while creating an Hazelcast instance to find and load the required class. 

Imagine that there are also other bundles and each of them has its own classloader. In this case, as a cache provider, Hazelcast doesn't responsible for finding all required classes.

Another workaround is as you mentioned is setting the context classloader before calling cache and reverting it then.

**Comment by exmortal** [comment]
You're right in a case when I use hazelcast directly, but I want to use it via JSR107 api. 
And the main goal of this issue is why bundle2ClassLoader is not using for deserialization for caches created using this manager ? 
```java
CacheManager manager = provider.getCacheManager(bundle2ClassLoader, null);
```
Maybe it's misunderstanding to JSR107 specification or my vision is not right.

**Comment by serkan-ozal** [comment]
JCache specification doesn't enforce specified classloader to be used as internal classloader, it is there for specifying scopes of `CacheManager`s.

And in your architecture, there maybe multiple users of the Hazelcast bundle and each one has its own classloader. Each one tries to use same cache with different classloader. This is not feasible from the cache provider perspective (Hazelcast)

**Comment by exmortal** [comment]
Could you please describe what does CacheManager's scopes means? (in a Hazelcast context)

JSR says: "To acquire non-default or alternative configurations of CacheManagers, for example with **custom
ClassLoaders** or caching implementor properties, developers should use one of the overloaded
CachingProvider getCacheManager methods."
And nothing said about the purpose of this "custom classloader". So I think it can be used for deserialization.

**Comment by serkan-ozal** [comment]
> Could you please describe what does CacheManager's scopes means? (in a Hazelcast context)

There can be multiple `CacheManager`s owned by same `CachingProvider` (`HazelcastCachingProvider`). Each of the `CacheManager` has their own classloader and URI so eventhough the cache with same name is created over these `CacheManager`s, in fact they are different caches. I mean there can be multiple caches with same name owned by different `CacheManager`s. This is considered as the scope of `CacheManager`. 

> JSR says: "To acquire non-default or alternative configurations of CacheManagers, for example with custom ClassLoaders or caching implementor properties, developers should use one of the overloaded
CachingProvider getCacheManager methods."
And nothing said about the purpose of this "custom classloader". So I think it can be used for deserialization.

Yes, there is no explicit definition for usage of classloader. Yes it can be used for deserialization but as you said, it **CAN**. But in your case, there can be multiple users/bundles of same the cache from the Hazelcast bundle and each of them may have their own classloader. How do you think it can be handled. Please don't think as there is only one Hazelcast cache owner bundle and only one cache user bundle. The API should be available for general purpose and I think this is one of the reason why there is no explicit definition about the usage of classloader especially in disributed environments.

**Comment by exmortal** [comment]
> Hazelcast bundle and each of them may have their own classloader. How do you think it can be handled. 
My idea was that each bundle has own cache manager created with bundle's classloader. 
Cache manager is available via custom API that restricts external bundles to get access to cache managers form other bundles. So that why I create this issue. 

Anyway, great thanks for a detailed answer, I think issue could be closed.

**Comment by serkan-ozal** [comment]
Ok thanks. I will keep this issue open a little bit to have a look what can be done. But I advice using thread context classloader as workaround at the moment.

**Comment by fuemf** [comment]
I ran in the same issue. It's a pity there is no fix :-(

I implemented a ServiceFactory<CacheManager> to get a custom CacheManager for every bundle that requests one. In the ServiceFactory I use the requesting bundle's ClassLoader to create the CacheManager.

In the java doc of CachingProvider.getCacheManager(..) I read:
> Requests a CacheManager [...] that uses the provided ClassLoader for loading underlying classes. 
So I thought it would be possible.

It took me some debugging to find out, that the ClassLoader of the SerializationService (which is the same for the whole hazelcast instance) instead.

Seems I have to find some ugly workaround.
